parsel/selector.py

```diff
@@ -267,34 +282,99 @@ def __init__(
 
-        if text is not None:
-            if not isinstance(text, str):
-                msg = f"text argument should be of type str, got {text.__class__}"
-                raise TypeError(msg)
-            root = self._get_root(text, base_url)
-        elif root is None:
+        self._text = text
+
+        if text is None and root is _NOTSET:
             raise ValueError("Selector needs either text or root argument")
 
+        if text is not None and not isinstance(text, str):
+            msg = f"text argument should be of type str, got {text.__class__}"
+            raise TypeError(msg)
+
+        if text is not None:
+            if type in ("html", "xml", None):
+                self._load_lxml_root(
+                    text, type=type or "html", base_url=base_url
+                )
+            elif type == "json":
+                self.root = _load_json_or_none(text)
+                self.type = type
+            else:
+                self.root = text
+                self.type = type
+        else:
+            self.root = root
+            if type is None and isinstance(self.root, etree._Element):
+                type = "html"
+            self.type = type or "json"
+
+        self._expr = _expr
         self.namespaces = dict(self._default_namespaces)
         if namespaces is not None:
             self.namespaces.update(namespaces)
-        self.root = root
-        self._expr = _expr
+
+    def _load_lxml_root(self, text, type, base_url=None):
+        self.type = type
+        self.root = self._get_root(text, base_url)
 
     def __getstate__(self) -> Any:
         raise TypeError("can't pickle Selector objects")
 
     def _get_root(self, text: str, base_url: Optional[str] = None) -> Any:
-        return create_root_node(text, self._parser, base_url=base_url)
+        return create_root_node(
+            text,
+            _ctgroup[self.type]["_parser"],
+            base_url=base_url,
+        )
+
+    def jmespath(
+        self: _SelectorType,
+        query: str,
+        **kwargs,
+    ) -> SelectorList[_SelectorType]:
+        """
+        Find objects matching the JMESPath ``query`` and return the result as a
+        :class:`SelectorList` instance with all elements flattened. List
+        elements implement :class:`Selector` interface too.
+
+        ``query`` is a string containing the `JMESPath
+        <https://jmespath.org/>`_ query to apply.
+
+        Any additional named arguments are passed to the underlying
+        ``jmespath.search`` call, e.g.::
+
+            selector.jmespath('author.name', options=jmespath.Options(dict_cls=collections.OrderedDict))
+        """
+        if self.type == "json":
+            data = self.root
+        elif isinstance(self.root, str):
+            data = _load_json_or_none(self.root)
+        elif self.root.text is None:
+            data = _load_json_or_none(self._text)
+        else:
+            data = _load_json_or_none(self.root.text)
```

Comment:
Maybe it's because it's evening here, but it's not very easy for me to understand this code :) Do you think it's possible to refactor it, or maybe add some comments? My current understanding - please correct me if that's wrong:

1. First condition is for json selectors, where self.root is data.
2. Second condition is for "text" selectors, i.e. results of some jmeselect runs? Can we check the type again (self.type = "text") to make it more clear, or is it problematic?
3. Third and forth conditions handle a case with trees; it seems there is something abuout lazy parsing going on here, but I'm not sure I understand when do we have case 3, and when it's case 4.

Reply:
Case 3 happens when you create a Selector with JSON data as `text` attribute and uses `jmespath`. It ends up having type equal to `"html"` and root as `HtmlElement` with no `text` attribute. It checks for `self.type` (which is `"html"`), then `self.root` (which is `HtmlElement`), and then checks for `self.root.text` (which is `None`). `self._text` is JSON so it loads and goes on. 

This can be verified in the first assert of _tests.test_selector_jmespath.JMESPathTestCase.test_json_has_html_ (line 37)

Case 4 happens when you use a Selector with HTML data as 'text' and uses  `jmespath`.
At this point, Selector has `type = 'html'`, root is an `HtmlElement` and `self.root.text` is a `str`.

This can be verified in the second assert of _tests.test_selector_jmespath.JMESPathTestCase.test_html_has_json_ (line 90)

My verdict: `type` looks messy.

Reply:
```
if self.type == "json":
    data = self.root
elif isinstance(self.root, str):
    data = _load_json_or_none(self.root)
else:
    data = _load_json_or_none(self.root.text or self._text) 
```
Could be a simple refactor. Everything is working fine, but it does feel that `type` was once a simple support check, but as a new `type` was added it raised some unnecessary complexity to the `Selector` which could probably just be ignored for the time being by standardizing it by first removing the extra `json` type and later on thinking of removing the attribute altogether and dealing with attribute types to handle the parsing alone.

Reply:
For example. All current `type = 'json'` could be tweaked by `isinstance(self.root, dict)`. Or, if other types may come (although with the current test suite no tests reach this 'if' statement with something else than a `dict`), such as lists or raw types, we could do a negative check such as `not isinstance(self.root, str) and self.root`

Then, this snippet could be refactored to:
```
if isinstance(self.root, str):
    data = _load_json_or_none(self.root)
elif isinstance(self.root, dict):
    data = self.root
else:
    data = _load_json_or_none(self.root.text or self._text)
```

Reply:
I think we should support JSON arrays at least.

It would be nice to support additional types, but I cannot think of a reason to use JMESPath on a simply JSON type like a string or a number, so I think it is OK not to support those if supporting them would complicate the user-facing API (i.e. require `type` where it could be omitted otherwise) or the implementation.

Reply:
I have simplified the code, it should be more readable now.